### PR TITLE
[1.19] Improve extensibility of DetectorRailBlock and PoweredRailBlock

### PR DESCRIPTION
**Pull request by XFactHD**
This PR improves the extensibility of the `DetectorRailBlock` and `PoweredRailBlock` by:
1. Allowing the implementor to pass in the blockstate property and associated value for the default state
2. Using the shape property getter to get the shape in `PoweredRailBlock#findPoweredRailSignal()`
3. Using the correct block to get the rail direction from in `PoweredRailBlock#isSameRailWithPower()`.

The main use case for this is something like a rail slope block with "underfill" that can only use ascending shapes and therefore should use a custom `EnumProperty<RailShape>`.

### CODE REVIEW

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
I think if this PR is going to get accepted, there needs to be more sanity checks in place such that providing a new shape property throws exceptions if certain methods aren't overridden. A custom shape property would essentially crash a good portion of the rail code.

The getter should be forced to throw an error if not overridden and patched into all instances in the class besides rotate and mirror, since we can just leave a note that these should be covered (issues for these are really only seen in templates).

**Comment by XFactHD** [comment]
> The getter should be forced to throw an error if not overridden

I had thought about just using the getter instead of a constructor parameter for the property, which would alleviate most of your concerns but this would lead to being unable to use non-static variables in the getter, which would potentially be a breaking change.
Would a `customProperty` flag used to throw in the getter be an acceptable solution?

> A custom shape property would essentially crash a good portion of the rail code.

There is exactly one relevant unpatched use of the hard-coded shape property that would crash: `PoweredRailBlock#updateState()`. The only other use of the shape property apart from `rotate()` and `mirror()` - `createBlockStateDefinition()` - cannot use a constructor parameter since it's called from the super constructor and therefore must be overriden anyway.
I'll add the missing patch for `PoweredRailBlock#updateState()`.

> besides rotate and mirror, since we can just leave a note that these should be covered (issues for these are really only seen in templates)

I'll add a doc comment to the added constructors to mention the need to override the getter, `rotate()`, `mirror()`, `createBlockStateDefinition()` and, potentially, `IForgeBaseRailBlock#isValidRailShape()`.

**Comment by ChampionAsh5357** [comment]
> Would a customProperty flag used to throw in the getter be an acceptable solution?

Yeah, that's fine. It would be the same way we handle `FluidType`s.

Everything else you mentioned should alleviate the concerns.

File: patches/minecraft/net/minecraft/world/level/block/DetectorRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/DetectorRailBlock.java.patch]: This can be one line.

File: patches/minecraft/net/minecraft/world/level/block/DetectorRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/DetectorRailBlock.java.patch]: This can be one line.

File: patches/minecraft/net/minecraft/world/level/block/DetectorRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/DetectorRailBlock.java.patch]: This can be one line.

File: patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch]: This can be one line.

File: patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch]: This can be one line.

File: patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch]: This can be one line.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Mainly can make the throwables all one line. Also, I'm not sure if `apiNote` compiles with gradle javadocs unless it's specified as a custom tag, so it probably should just be made as the constructor comment.

**Review by ChampionAsh5357 — APPROVED** [review]

**Comment by ChampionAsh5357** [comment]
After some discussion, it boils down to these changes not really providing any benefit to the end modder.

First, in conjunction with the block state property, `RailShape` is not extensible. While you could use the `EnumProperty` as you mentioned to limit the subset of enum values allowed, that could also be managed with proper programming of the rails, which would need to be done anyways since anything that calls an incorrect shape would ultimately throw an error. It wouldn't really matter in creative mode since they can do everything anyways.

As for the `getRailDirection` change, it'll need some additional justification, and would probably only be merged in a later version as it constitutes a breaking change. From what I understand, the method is just delegating to the subsequent method as the rail direction should be determined by the block whose rail shape we are getting, which currently breaks the logic contract of the method itself. However, if you could clarify the key points and issues with its implementation currently, that would be good as well.

I will mark this as a WIP for 1.20 for these reasons.

**Comment by XFactHD** [comment]
> First, in conjunction with the block state property, `RailShape` is not extensible. While you could use the `EnumProperty` as you mentioned to limit the subset of enum values allowed, that could also be managed with proper programming of the rails, which would need to be done anyways since anything that calls an incorrect shape would ultimately throw an error. It wouldn't really matter in creative mode since they can do everything anyways.

I would argue against this that providing more values in a property than the respective block actually uses and "supports" is a bad implementation and is just asking for issues. Case-in-point: I have a block unrelated to this that has two properties that can produce "impossible" combinations which cause issues with mods that expect all possible property combinations to be valid (which is a completely valid assumption in my opinion).
In terms of causing issues due to setting invalid values: everything in vanilla code that sets rail shapes either must be overriden to not throw when a custom property is used (`rotate()` and `mirror()`) or is guarded by a call to `IForgeBaseRailBlock#isValidRailShape()` (connection checks in `RailState`), which is also mentioned in the constructor docs as potentially needing to be overriden.

&nbsp;
 
> As for the `getRailDirection` change, it'll need some additional justification, and would probably only be merged in a later version as it constitutes a breaking change. From what I understand, the method is just delegating to the subsequent method as the rail direction should be determined by the block whose rail shape we are getting, which currently breaks the logic contract of the method itself. However, if you could clarify the key points and issues with its implementation currently, that would be good as well.

The reason for that change is that at the moment `getRailDirection()` is called on the wrong block given the `BlockState` passed to it. `PoweredRailBlock#isSameRailWithPower()` is called on the block searching for powered neighbours and gets the `BlockPos` of a potential neighbor. That position is used to get the neighbouring state. When the neighbor is a `PoweredRailBlock`, then the neighbor's state is passed to `BaseRailBlock#getRailDirection()` which is called on the searching block, which leads to the **searching** block being used to determine the direction of the **searched** block, which in turn causes an exception when one of those blocks uses a non-standard shape property. The change makes sure that `getRailDirection()` is called on the **searched** block.
For this reason I would argue that this change is a bugfix and in no way a breaking change.

**Comment by ChampionAsh5357** [comment]
After further discussion, Lex came up with [this diff](https://gist.github.com/ChampionAsh5357/d135a276aab48a8781174682d4a5bda9). I would've uploaded the file itself; however GitHub doesn't like `.diff`, so I just made it a gist.

First, this addresses the default state issue by migrating the call to a separate protected method (aptly named `#registerDefaultState`) which can be overridden. No new constructors need to be added because of this.

Second, this replaces all simple calls to the field with the `#getShapeProperty` method. Note that this does not replace `#rotate` or `#mirror`; however, a separate comment was added on the `#getShapeProperty` indicating that other methods may break without proper inspection, so this will hopefully mitigate those issues.

Finally, `#getRailDirection` was corrected to look at the other rail in addition to replacing all the casts with pattern matching in the instance check.

If there are no other concerns, you can implement the following diff for your PR.

**Comment by XFactHD** [comment]
I have implemented the patches as suggested. The only concern I have with them is the use of `getShapeProperty()` in `createStateDefinition()` as mentioned earlier but I'll let the core team decide whether they consider this a valid concern:
> I had thought about just using the getter instead of a constructor parameter for the property, which would alleviate most of your concerns but this would lead to being unable to use non-static variables in the getter, which would potentially be a breaking change.

File: patches/minecraft/net/minecraft/world/level/block/BaseRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/BaseRailBlock.java.patch:42]: This is a regular comment and not a javadoc. This is to avoid conflicts with mapping sets (e.g. parchment) adding their own javadoc.

File: patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch:38]: Replacement of casting with pattern matching not added

File: patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch]: This should check the other state for finding the powered rail signal.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
> I have implemented the patches as suggested. The only concern I have with them is the use of getShapeProperty() in createStateDefinition() as mentioned earlier but I'll let the core team decide whether they consider this a valid concern

Fair enough, though I'd argue you shouldn't be supplying properties as instances.

**Comment by XFactHD** [comment]
Ups, missed the pattern change. Also good to know with the comment.

**Review by ChampionAsh5357 — APPROVED** [review]

File: patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/level/block/PoweredRailBlock.java.patch]: should be isActivatorRail function calls

File: src/main/resources/forge.exc
Comment (LexManos) [inline src/main/resources/forge.exc:82]: should not be needed

**Pull request merged**
